<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>The human genome reference &mdash; halvade 1.1.0 documentation</title>
    
    <link rel="stylesheet" href="../_static/classic.css" type="text/css" />
    <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../',
        VERSION:     '1.1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../_static/jquery.js"></script>
    <script type="text/javascript" src="../_static/underscore.js"></script>
    <script type="text/javascript" src="../_static/doctools.js"></script>
    <link rel="top" title="halvade 1.1.0 documentation" href="../index.html" />
    <link rel="next" title="Hadoop setup" href="hadoop.html" />
    <link rel="prev" title="Installation" href="installation.html" /> 
  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="hadoop.html" title="Hadoop setup"
             accesskey="N">next</a> |</li>
        <li class="right" >
          <a href="installation.html" title="Installation"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">halvade 1.1.0 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="the-human-genome-reference">
<h1>The human genome reference<a class="headerlink" href="#the-human-genome-reference" title="Permalink to this headline">¶</a></h1>
<p>Halvade uses the genome reference fasta file (<code class="docutils literal"><span class="pre">ucsc.hg19.fasta</span></code>) found in the GATK resource bundle to build the index files for both BWA and STAR. The fasta file comes together with an index and a dictionary file. Additionally a full dbsnp file (version 138) is used when recalibrating the basescores for the reads. These files are all found in the GATK resource bundle which is available <a class="reference external" href="ftp://gsapubftp-anonymous&#64;ftp.broadinstitute.org/bundle/2.8/hg19/">here</a>. The ftp site has a limited number of parallel downloads and might not load at these times. Here is how you download the files using the terminal:</p>
<div class="highlight-bash"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3
4
5</pre></div></td><td class="code"><div class="highlight"><pre><span></span>wget ftp://gsapubftp-anonymous@ftp.broadinstitute.org/bundle/2.8/hg19/ucsc.hg19.fasta.gz
wget ftp://gsapubftp-anonymous@ftp.broadinstitute.org/bundle/2.8/hg19/ucsc.hg19.fasta.fai.gz
wget ftp://gsapubftp-anonymous@ftp.broadinstitute.org/bundle/2.8/hg19/ucsc.hg19.dict.gz
wget ftp://gsapubftp-anonymous@ftp.broadinstitute.org/bundle/2.8/hg19/dbsnp_138.hg19.vcf.gz
wget ftp://gsapubftp-anonymous@ftp.broadinstitute.org/bundle/2.8/hg19/dbsnp_138.hg19.vcf.idx.gz
</pre></div>
</td></tr></table></div>
<p>Next we need to unzip all these files so they can be used in Halvade:</p>
<div class="highlight-bash"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2</pre></div></td><td class="code"><div class="highlight"><pre><span></span>gunzip ucsc.hg19.fasta.gz <span class="o">&amp;&amp;</span> gunzip ucsc.hg19.fasta.fai.gz <span class="o">&amp;&amp;</span> gunzip ucsc.hg19.dict.gz <span class="o">&amp;&amp;</span> <span class="se">\</span>
gunzip dbsnp_138.hg19.vcf.gz <span class="o">&amp;&amp;</span> gunzip dbsnp_138.hg19.vcf.idx.gz
</pre></div>
</td></tr></table></div>
<p>The index files, the reference and the dbsnp file need to be uploaded to the HDFS server if a cluster is used to run Halvade. Setting up Halvade is described in the next part of the documentation.</p>
<div class="section" id="bwa-reference-for-wgs-wes-data">
<h2>BWA reference for WGS/WES data<a class="headerlink" href="#bwa-reference-for-wgs-wes-data" title="Permalink to this headline">¶</a></h2>
<p>For the whole genome and exome sequencing pipelines the BWA aligner is used. To run BWA an index needs to be created of the reference fasta file, which will needs to be accessible by Halvade so BWA can be called correctly. The BWA binary is available in the <code class="docutils literal"><span class="pre">bin.tar.gz</span></code> archive, which is provided in every Halvade release. However, this is also needed by Halvade itself, so first make a copy of this archive and extract the copy.</p>
<div class="highlight-bash"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3</pre></div></td><td class="code"><div class="highlight"><pre><span></span>cp bin.tar.gz copy_bin.tar.gz
tar -xvf copy_bin.tar.gz
./bin/bwa index ucsc.hg19.fasta
</pre></div>
</td></tr></table></div>
<p>This process will create 5 files with the provided name as a prefix, the naming is important as Halvade finds this index by the given fasta prefix <code class="docutils literal"><span class="pre">ucsc.hg19</span></code>.</p>
</div>
<div class="section" id="star-reference-for-rna-seq-data">
<h2>STAR reference for RNA-seq data<a class="headerlink" href="#star-reference-for-rna-seq-data" title="Permalink to this headline">¶</a></h2>
<p>The RNA-seq pipeline uses the STAR aligner to perform the read alignment step. Similarly to BWA, the STAR aligner requires an index of the reference fasta file. Again this can be created by using the STAR binary which is provided in the <code class="docutils literal"><span class="pre">bin.tar.gz</span></code> archive which is available in every Halvade release. Make a copy of this archive as it will be needed by Halvade later on. This process can be done in the terminal as follows:</p>
<div class="highlight-bash"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3</pre></div></td><td class="code"><div class="highlight"><pre><span></span>cp bin.tar.gz copy_bin.tar.gz
tar -xvf copy_bin.tar.gz
./bin/STAR .............
</pre></div>
</td></tr></table></div>
<p>After the STAR genome index has been created, the provided output folder will contain all files needed by STAR and in turn Halvade.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="../index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">The human genome reference</a><ul>
<li><a class="reference internal" href="#bwa-reference-for-wgs-wes-data">BWA reference for WGS/WES data</a></li>
<li><a class="reference internal" href="#star-reference-for-rna-seq-data">STAR reference for RNA-seq data</a></li>
</ul>
</li>
</ul>

  <h4>Previous topic</h4>
  <p class="topless"><a href="installation.html"
                        title="previous chapter">Installation</a></p>
  <h4>Next topic</h4>
  <p class="topless"><a href="hadoop.html"
                        title="next chapter">Hadoop setup</a></p>
  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="../_sources/contents/references.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="../search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="../genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="hadoop.html" title="Hadoop setup"
             >next</a> |</li>
        <li class="right" >
          <a href="installation.html" title="Installation"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="../index.html">halvade 1.1.0 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &copy; Copyright 2016, Dries Decap.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.1.
    </div>
  </body>
</html>